virtool/account/data.py

```diff
@@ -0,0 +1,364 @@
+
+        :param user_id: the user ID
+        :param key_id: the ID of the API key to get
+        :param api_key_projection: the projection for the api key
+        :return: the API key
+        """
+        document = await self._db.keys.find_one(
+            {"id": key_id, "user.id": user_id}, api_key_projection
+        )
+
+        if document is None:
+            raise ResourceNotFoundError()
+
+        return document
+
+    async def edit_key(
+        self,
+        user_id: str,
+        key_id: str,
+        data: EditKeySchema,
+        api_key_projection: dict,
+    ) -> dict:
+        """
+        Change the permissions for an existing API key.
+
+        :param user_id: the user ID
+        :param key_id: the ID of the API key to update
+        :param data: permissions update
+        :param api_key_projection: the projection for the api key
+        :return: the API key
+        """
+        if data.permissions is None:
+            permissions_update = {}
+        else:
+            permissions_update = data.permissions.dict(exclude_unset=True)
+
+        if not await self._db.keys.count_documents({"id": key_id}):
+            raise ResourceNotFoundError()
+
+        user = await self._db.users.find_one(user_id, ["administrator", "permissions"])
+
+        # The permissions currently assigned to the API key.
+        permissions = await get_one_field(
+            self._db.keys, "permissions", {"id": key_id, "user.id": user_id}
+        )
+
+        permissions.update(permissions_update)
+
+        if not user["administrator"]:
+            permissions = limit_permissions(permissions, user["permissions"])
+
+        document = await self._db.keys.find_one_and_update(
+            {"id": key_id},
+            {"$set": {"permissions": permissions}},
+            projection=api_key_projection,
+        )
+
+        return document
+
+    async def delete_key(self, user_id: str, key_id: str):
+        """
+        Delete an API key by its id.
+
+        :param user_id: the user ID
+        :param key_id: the ID of the API key to delete
+        """
+        delete_result = await self._db.keys.delete_one(
+            {"id": key_id, "user.id": user_id}
+        )
+
+        if delete_result.deleted_count == 0:
+            raise ResourceNotFoundError()
+
+    async def login(self, session_id: str, data: CreateLoginSchema) -> Union[int, str]:
+        """
+        Create a new session for the user with `username`.
+
+        :param session_id: the login session ID
+        :param data: the login data
+        """
+        # When `remember` is set, the session will last for 1 month instead of the
+        # 1-hour default
+
+        # Re-render the login page with an error message if the username doesn't
+        # correlate to a user_id value in
+        # the database and/or password are invalid.
+        handle = data.username
+        password = data.password
+        remember = data.remember
```

Comment:
Not necessary to redeclare attributes as separate variables. This was done when using dictionaries to improve readibility.